go/src/meguca/server/embed.go

```diff
@@ -12,152 +11,142 @@ import (
-			return errYouTubeLive(ytid)
-		}
-
-		vidFormats := info.Formats.
-			Filter(ytdl.FormatExtensionKey, []interface{}{"webm"}).
-			Filter(ytdl.FormatResolutionKey, []interface{}{"360p"}).
-			Filter(ytdl.FormatAudioEncodingKey, []interface{}{"aac", "opus", "vorbis"}).
-			Best(ytdl.FormatVideoEncodingKey)
-			
-		if len(vidFormats) == 0 {
-			vidFormats = info.Formats.
-				Filter(ytdl.FormatExtensionKey, []interface{}{"webm"}).
-				Filter(ytdl.FormatResolutionKey, []interface{}{"144p", "240p", "270p", "360p"}).
-				Filter(ytdl.FormatAudioEncodingKey, []interface{}{"aac", "opus", "vorbis"}).
-				Best(ytdl.FormatResolutionKey).
-				Best(ytdl.FormatVideoEncodingKey)
+	info, err := getYouTubeInfo(extractParam(r, "id"))
 
-			if len(vidFormats) == 0 {
-				vidFormats = info.Formats.
-					Filter(ytdl.FormatExtensionKey, []interface{}{"mp4", "flv", "3gp", "ts"}).
-					Filter(ytdl.FormatResolutionKey, []interface{}{"144p", "240p", "270p", "360p"}).
-					Filter(ytdl.FormatAudioEncodingKey, []interface{}{"aac", "opus", "vorbis"}).
-					Best(ytdl.FormatResolutionKey).
-					Best(ytdl.FormatVideoEncodingKey)
-				
-				if len(vidFormats) == 0 {
-					return errNoYoutubeVideo(ytid)
-				}
-			}
-		}
+	if err != nil {
+		httpError(w, r, err)
+		return
+	}
 
-		video, err := info.GetDownloadURL(vidFormats[0])
+	w.Write([]byte(info))
+}
 
-		if err != nil {
-			return
-		}
+// Get BitChute video title by ID
+func bitChuteTitle(w http.ResponseWriter, r *http.Request) {
+	s, err := goscraper.Scrape("https://www.bitchute.com/embed/" + extractParam(r, "id"), 3)
 
-		// Unfortunately, in some cases you cannot get 720p with only webm
-		vidFormats = info.Formats.
-			Filter(ytdl.FormatAudioEncodingKey, []interface{}{"aac", "opus", "vorbis"}).
-			Best(ytdl.FormatResolutionKey)
+	if err != nil {
+		httpError(w, r, err)
+		return
+	}
 
-		if len(vidFormats) == 0 {
-			return errNoYoutubeVideo(ytid)
-		}
+	w.Write([]byte(s.Preview.Description))
+}
 
-		videoHigh, err := info.GetDownloadURL(vidFormats[0])
+// Return YouTube video info and handle errors
+func getYouTubeInfo(id string) (ret string, err error) {
+	var ok bool
+	info, err := ytdl.GetVideoInfoFromID(id)
 
-		if err != nil {
-			return
-		}
+	if err != nil {
+		return ret, errYouTube(id, err)
+	} else if info.Duration == 0 {
+		return ret, errYouTubeLive(id)
+	}
 
-		thumb := info.GetThumbnailURL(ytdl.ThumbnailQualityMaxRes)
+	thumb := info.GetThumbnailURL(ytdl.ThumbnailQualityMaxRes)
 
-		for i := 0; i < 5; i++ {
-			ok, err := func() (bool, error) {
-				// Perhaps there is a way to check the status code without fetching the body?
-				resp, err := http.Get(thumb.String())
+	for _, val := range [4]ytdl.ThumbnailQuality {
+		ytdl.ThumbnailQualityHigh,
+		ytdl.ThumbnailQualityMedium,
+		ytdl.ThumbnailQualityDefault,
+		ytdl.ThumbnailQualitySD,
+	} {
+		resp, err := http.Get(thumb.String())
 
-				if err != nil {
-					return false, err
-				}
+		if err == nil {
+			defer resp.Body.Close()
```

Comment:
Is this needed if we're only checking for the status code?

Reply:
Yes. You'd leak memory otherwise. You should also close the body right after the call, not only on error. No need to defer as well. Do I need to explain why?

Reply:
No need, it makes sense.
I didn't realize the body opens right at initialization.